crates/holochain/src/core/workflow/countersigning_workflow/incomplete.rs

```diff
@@ -0,0 +1,332 @@
+use crate::conductor::space::Space;
+use crate::core::queue_consumer::TriggerSender;
+use crate::core::workflow::countersigning_workflow::{
+    success, SessionCompletionDecision, SessionResolutionOutcome, NUM_AUTHORITIES_TO_QUERY,
+};
+use crate::core::workflow::{countersigning_workflow, WorkflowResult};
+use crate::prelude::{Entry, RecordEntry};
+use either::Either;
+use holo_hash::AgentPubKey;
+use holochain_cascade::CascadeImpl;
+use holochain_p2p::actor::GetActivityOptions;
+use holochain_p2p::HolochainP2pDnaT;
+use holochain_state::prelude::{
+    current_countersigning_session, CurrentCountersigningSessionOpt, SourceChainResult,
+};
+use holochain_types::activity::ChainItems;
+use holochain_zome_types::entry::GetOptions;
+use holochain_zome_types::prelude::{
+    ChainQueryFilter, ChainQueryFilterRange, ChainStatus, SignedAction,
+};
+use itertools::Itertools;
+use std::sync::Arc;
+
+/// Resolve an incomplete countersigning session.
+///
+/// This function is responsible for deciding what action to take when a countersigning session
+/// has failed to complete.
+///
+/// The function returns true if the session state has been cleaned up and the chain has been unlocked.
+/// Otherwise, the function returns false and the cleanup needs to be retried to resolved manually.
+pub async fn inner_countersigning_session_incomplete(
+    space: Space,
+    network: Arc<impl HolochainP2pDnaT>,
+    author: AgentPubKey,
+) -> WorkflowResult<(SessionCompletionDecision, Vec<SessionResolutionOutcome>)> {
+    let authored_db = space.get_or_create_authored_db(author.clone())?;
+
+    let maybe_current_session = authored_db
+        .write_async({
+            let author = author.clone();
+            move |txn| -> SourceChainResult<CurrentCountersigningSessionOpt> {
+                let maybe_current_session =
+                    current_countersigning_session(txn, Arc::new(author.clone()))?;
+
+                Ok(maybe_current_session)
+            }
+        })
+        .await?;
+
+    if maybe_current_session.is_none() {
+        tracing::error!("Countersigning session was in an unknown state but no session entry was found. Holochain is only meant to enter this state when there is an entry to remove and won't recover: {:?}", author);
+        return Ok((
+            SessionCompletionDecision::Indeterminate,
+            Vec::with_capacity(0),
+        ));
+    }
+
+    // Now things get more complicated. We have a countersigning entry on our chain but the session
+    // is in a bad state. We need to figure out what the session state is and how to resolve it.
```

Comment:
Should Holochain really try to figure that out? It's great that there is a way to do it, but it's also super complex.

Reply:
You are correct, it's complex. It is also technically something that you could achieve yourself using a client. You'd struggle right now because the interface for agent activity sucks but it could be done. I think the benefit here is just taking the number of times you have to interact with this and understand it as a user down as low as possible...

It's a fair question though, and something I'm happy to discuss the pros and cons of

Reply:
Generally I feel the approach is to leave such operations to a happ developer and that Holochain does not come with such complexity. But since you've figured this out already, I think we can keep it in and provide this improved experience where one participant abandoning a session after a cs entry has been committed doesn't require happ dev intervention to resolve. It may outweigh the difficulty to debug and fix bugs in this part of the code. I'm only sure of the benefit right now 😀